sdk/storage/azure-storage-blob-nio/src/test/java/com/azure/storage/blob/nio/AzureFileSystemProviderSpec.groovy

```diff
@@ -111,4 +114,187 @@ class AzureFileSystemProviderSpec extends APISpec {
         expect:
         provider.getScheme() == "azb"
     }
+
+    @Unroll
+    def "FileSystemProvider createDir parent exists"() {
+        setup:
+        def fs = createFS(config)
+
+        // Generate resource names.
+        // Don't use default directory to ensure we honor the root.
+        def rootName = fs.getRootDirectories().last().toString()
+        def containerName = rootName.substring(0, rootName.length() - 1)
+        def parent = ""
+        for (int i=0; i < depth; i++) {
+            parent += generateBlobName() + AzureFileSystem.PATH_SEPARATOR
+        }
+        def dirName = generateBlobName()
+        def dirPathStr = parent + dirName
+
+        def dirPath = fs.getPath(rootName, dirPathStr)
+
+        // Generate clients to resources. Create resources as necessary
+        def containerClient = primaryBlobServiceClient
+            .getBlobContainerClient(containerName)
+        /*
+        In this case, we are putting the blob in the root directory, i.e. directly in the container, so no need to
+        create a blob.
+         */
+        if (parent != "") {
+            def parentClient = containerClient.getBlobClient(parent)
+            parentClient.getAppendBlobClient().create()
+        }
+        def dirClient = containerClient.getBlobClient(dirPathStr)
+
+        when:
+        fs.provider().createDirectory(dirPath)
+
+        then:
+        dirClient.getPropertiesWithResponse(null, null, null).getValue().getMetadata()
+            .containsKey(AzureFileSystemProvider.DIR_METADATA_MARKER)
+
+        where:
+        depth | _
+        0     | _ // Test putting a blob in the root dir.
+        1     | _
+        2     | _
+    }
+
+    def "FileSystemProvider createDir relativePath"() {
+        setup:
+        def fs = createFS(config)
+        def containerClient =
+            primaryBlobServiceClient.getBlobContainerClient(rootToContainer(fs.getDefaultDirectory().toString()))
+        AppendBlobClient blobClient = containerClient.getBlobClient("foo").getAppendBlobClient()
+
+        when: "Relative paths are resolved against the default directory"
+        fs.provider().createDirectory(fs.getPath("foo"))
+
+        then:
+        blobClient.getProperties().getMetadata().containsKey(AzureFileSystemProvider.DIR_METADATA_MARKER)
+    }
+
+    def "FileSystemProvider createDir file already exists"() {
+        setup:
+        def fs = createFS(config)
+        def containerClient =
+            primaryBlobServiceClient.getBlobContainerClient(rootToContainer(fs.getDefaultDirectory().toString()))
+        AppendBlobClient blobClient = containerClient.getBlobClient("foo").getAppendBlobClient()
+
+        when: "A file already exists at the location"
```

Comment:
This is a bit of a departure from how we normally do tests. It seemed fitting to me, but if others feel I should break these up into discrete tests, I can do that, too.

Reply:
I think using the descriptions of the whens help make it clearer, I have no strong oppositions about this

Reply:
I'd prefer separation, as it's not clear which when/then fails when one does. Additionally, we'll stop when we encounter a single failure and we won't immediately see which specific one failed, which is useful information when initially looking at a failure.

If there is a way to have similar mechanics as table tests with `@Unroll` added, then I think this would be useful, as setup code can be complex sometimes.